cmd/car/compile.go

```diff
@@ -0,0 +1,446 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"unicode/utf8"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	carv1 "github.com/ipld/go-car"
+	"github.com/ipld/go-car/util"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/ipld/go-ipld-prime/storage/memstore"
+	"github.com/urfave/cli/v2"
+)
+
+// Compile is a command to translate between a human-debuggable patch-like format and a car file.
+func CompileCar(c *cli.Context) error {
+	var err error
+	inStream := os.Stdin
+	if c.Args().Len() >= 1 {
+		inStream, err = os.Open(c.Args().First())
+		if err != nil {
+			return err
+		}
+	}
+
+	//parse headers.
+	br := bufio.NewReader(inStream)
+	header, _, err := br.ReadLine()
+	if err != nil {
+		return err
+	}
+
+	v2 := strings.Contains(string(header), "--v2")
+	trimH := strings.TrimSpace(string(header))
+	headerParts := strings.Split(trimH, " ")
+	carName := headerParts[len(headerParts)-1]
+
+	roots := make([]cid.Cid, 0)
+	for {
+		peek, err := br.Peek(4)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		if bytes.Equal(peek, []byte("--- ")) {
+			break
+		}
+		rootLine, _, err := br.ReadLine()
+		if err != nil {
+			return err
+		}
+		if strings.HasPrefix(string(rootLine), "root ") {
+			var rCidS string
+			fmt.Sscanf(string(rootLine), "root %s", &rCidS)
+			rCid, err := cid.Parse(rCidS)
+			if err != nil {
+				return err
+			}
+			roots = append(roots, rCid)
+		}
+	}
+
+	//parse blocks.
+	rawBlocks := make(map[cid.Cid][]byte)
+	rawCodecs := make(map[cid.Cid]string)
+
+	for {
+		nextCid, mode, nextBlk, err := parsePatch(br)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		rawBlocks[nextCid] = nextBlk
+		rawCodecs[nextCid] = mode
+	}
+
+	//fmt.Printf("structuring as tree...\n")
+	// structure as a tree
+	childMap := make(map[cid.Cid][]cid.Cid)
```

Comment:
whoa, so I think this whole next section (x2) exists because we're not confident that the resulting CIDs will match the input CIDs, so we're doing a search, reconstruct, replace operation on them all, is that right?

  1. why do we not have confidence they're going to reconstruct byte-perfect, shouldn't that be reasonable? There might be a CIDv0 CIDv1 difference but we could easily do that check -- if expected CID is v0 then downcast actual CID and compare.
  2. why don't we just error if the resulting CID of the reconstructed block doesn't match the expected?

Reply:
Thought about this after writing and I realise that it's probably because we anticipate some input to come in badly encoded forms so round-tripping is going to result in mismatched CIDs. So I guess that's why this is here.

This does seem like a lot of effort to go to though; and it's also a little error-prone just replacing the CIDs as strings. That won't necessarily always find the actual links, they could just be included as text and it assumes that they want them to be changed. A find and replace should probably be at least looking for `"/": "...."\n`, but even that's not necessarily accurate either. A more complete approach would be to walk the instantiated data model form and change the links out, but I guess that's even more code complication! Then there's the CIDv0 vs CIDv1 thing, what if the original wants to be in CIDv0 but we up-convert them to CIDv1?

Lots of effort to do it the _right_ way, but it makes me question whether doing it at all is worth it.

Your call I guess, but maybe add some comments in here about what it's doing so the first person to encounter a bug with this knows why and can choose whether to fix it. It wasn't clear to me what it was doing until I walked through the whole two blocks.

Reply:
this is for the 'i want to mutate the graph and re-build a valid tree' case.
if i change the json in the patch, then that block will have a new cid when it's hashed.
that hash change needs to propagate back up the DAG.

data model still won't get all edge cases - what if it's a different codec for the same MH? what if the link is encoded in a block that's raw or that we can't parse?

this is the first pass that works reasonably well for the json/cbor cases that i've attempted to manually edit.

i am unconvinced it's worth the time to do much more complex work and still not do something perfect vs where it is currently, which supports a pretty valuable use case already.

Reply:
oh right, tinker with the content and regenerate the graph; fair enough, I can see that being useful - it does seem like something that needs clear caveats in comments though!